storage/grpc_client.go

```diff
@@ -964,3 +1029,301 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
 	r.cancel = cancel
 	return res.response, nil
 }
+
+func newGRPCWriter(c *grpcStorageClient, params *openWriterParams, r io.Reader) *gRPCWriter {
+	size := params.chunkSize
+	if params.chunkSize == 0 {
+		// TODO: Should we actually use the minimum of 256 KB here when the user
+		// indicates they want minimal memory usage? We cannot do a zero-copy,
+		// bufferless upload like HTTP/JSON can.
+		// TODO: We need to determine if we can avoid starting a
+		// resumable upload when the user *plans* to send more than bufSize but
+		// with a bufferless upload.
+		size = maxPerMessageWriteSize
+	}
+
+	return &gRPCWriter{
+		buf:           make([]byte, size),
+		c:             c,
+		ctx:           params.ctx,
+		reader:        r,
+		bucket:        params.bucket,
+		object:        params.object,
+		attrs:         params.attrs,
+		conds:         params.conds,
+		encryptionKey: params.encryptionKey,
+		sendCRC32C:    params.sendCRC32C,
+	}
+}
+
+// gRPCWriter is a wrapper around the the gRPC client-stream API that manages
+// sending chunks of data provided by the user over the stream.
+type gRPCWriter struct {
+	c      *grpcStorageClient
+	buf    []byte
+	reader io.Reader
+
+	ctx context.Context
+
+	bucket        string
+	object        string
+	attrs         *ObjectAttrs
+	conds         *Conditions
+	encryptionKey []byte
+
+	sendCRC32C bool
+
+	// The gRPC client-stream used for sending buffers.
+	stream storagepb.Storage_WriteObjectClient
+
+	// The Resumable Upload ID started by a gRPC-based Writer.
+	upid string
+}
+
+// startResumableUpload initializes a Resumable Upload with gRPC and sets the
+// upload ID on the Writer.
+//
+// This is an experimental API and not intended for public use.
```

Comment:
Can probably remove the experimental tags since it would apply to everything in this file.

Reply:
Good call. Done.